fs2-google-pubsub-grpc/src/main/scala/com/permutive/pubsub/producer/grpc/internal/DefaultPublisher.scala

```diff
@@ -41,14 +41,18 @@ private[pubsub] class DefaultPublisher[F[_], A: MessageEncoder](
         for {
           future <- F.delay(publisher.publish(message))
           result <- F.async[MessageId] { cb =>
-            ApiFutures.addCallback(
-              future,
-              new ApiFutureCallback[String] {
-                override def onFailure(t: Throwable): Unit   = cb(Left(t))
-                override def onSuccess(result: String): Unit = cb(Right(MessageId(result)))
-              },
-              callbackExecutor
-            )
+            F.defer {
+              ApiFutures.addCallback(
+                future,
+                new ApiFutureCallback[String] {
+                  override def onFailure(t: Throwable): Unit   = cb(Left(t))
+                  override def onSuccess(result: String): Unit = cb(Right(MessageId(result)))
+                },
+                callbackExecutor
+              )
+
+              F.delay(Option(F.blocking(future.cancel(true)).void))
```

Comment:
Not sure about this

Reply:
I think I see what you're trying to do from an example in the CE docs. I think I might rewrite it like so, just so you can be sure the `addCallback` has been properly suspended:

```scala
F.delay {
           ApiFutures.addCallback(
              new ApiFutureCallback[String] {
                future,
                override def onFailure(t: Throwable): Unit   = cb(Left(t))
                new ApiFutureCallback[String] {
                override def onSuccess(result: String): Unit = cb(Right(MessageId(result)))
                  override def onFailure(t: Throwable): Unit   = cb(Left(t))
              },
                  override def onSuccess(result: String): Unit = cb(Right(MessageId(result)))
              callbackExecutor
                },
            )
                callbackExecutor
              )

	Some(F.delay(future.cancel(true)).void))
}
```

I've put the CE3 examples put the future cancellation in a `delay` block, but that's for a scala future, so I don't know whether the google futures behave in the same, I should think they do in this regard.

Reply:
FWIW here's what we are using in Bigtable (`utils`) which had evolved a bit beyond here as it used `Concurrent[F].cancelable`: https://github.com/permutive/permutive-utils/blob/4845fd2368929b1a8bf0fa5c48a95fce1aec06ab/modules/google-bigtable/src/main/scala/com/permutive/google/bigtable/utils/FutureInterop.scala#L11

It also sets `mayInterruptIfRunning` to `false` as I found in testing for Bigtable that wasn't helpful. This may differ though...

I know your question was more around applying the cancellation we may want to grab this version seeing as it's more up-to-date...

Reply:
Discussed offline but I had totally forgotten `Concurrent` has changed (effectively) to `Async` 🤦